pkg/controllers/management/restrictedadminrbac/fleet.go

```diff
@@ -50,42 +51,72 @@ func (r *rbaccontroller) ensureRestricedAdminForFleet(key string, obj *v3.Global
 }
 
 func (r *rbaccontroller) ensureRolebinding(namespace string, subject k8srbac.Subject, grb *v3.GlobalRoleBinding) error {
-	rbName := fmt.Sprintf("%s-fleetworkspace-%s", grb.Name, rbac.RestrictedAdminClusterRoleBinding)
-	rb := &k8srbac.RoleBinding{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      rbName,
-			Namespace: namespace,
-			Labels:    map[string]string{rbac.RestrictedAdminClusterRoleBinding: "true"},
-			OwnerReferences: []metav1.OwnerReference{
-				{
-					APIVersion: grb.TypeMeta.APIVersion,
-					Kind:       grb.TypeMeta.Kind,
-					UID:        grb.UID,
-					Name:       grb.Name,
-				},
-			},
-		},
-		RoleRef: k8srbac.RoleRef{
-			Name: "fleetworkspace-admin",
-			Kind: "ClusterRole",
-		},
-		Subjects: []k8srbac.Subject{
-			subject,
+	name := fmt.Sprintf("%s-fleetworkspace-%s", grb.Name, rbac.RestrictedAdminClusterRoleBinding)
+	ownerRefs := []metav1.OwnerReference{
+		{
+			APIVersion: grb.TypeMeta.APIVersion,
+			Kind:       grb.TypeMeta.Kind,
+			UID:        grb.UID,
+			Name:       grb.Name,
 		},
 	}
-	_, err := r.rbLister.Get(namespace, rbName)
+	roleRef := k8srbac.RoleRef{
+		Name: "fleetworkspace-admin",
+		Kind: "ClusterRole",
+	}
+	subjects := []k8srbac.Subject{
+		subject,
+	}
+
+	rb, err := r.rbLister.Get(namespace, name)
 	if err != nil && !k8serrors.IsNotFound(err) {
+		// list call failed for unknown reason, give up
 		return err
-	} else if k8serrors.IsNotFound(err) {
-		_, err = r.roleBindings.Create(rb)
+	}
+
+	// role binding not found, create it
+	if err != nil && k8serrors.IsNotFound(err) {
+		_, err = r.roleBindings.Create(&k8srbac.RoleBinding{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:            name,
+				Namespace:       namespace,
+				Labels:          map[string]string{rbac.RestrictedAdminClusterRoleBinding: "true"},
+				OwnerReferences: ownerRefs,
+			},
+			RoleRef:  roleRef,
+			Subjects: subjects,
+		})
 		if err != nil && !k8serrors.IsAlreadyExists(err) {
 			return err
 		}
-	} else {
-		_, err := r.roleBindings.Update(rb)
-		if err != nil {
-			return err
-		}
+		return nil
+	}
+
+	// role binding found, possibly in dirty state. Make sure relevant fields
+	// are set right
+	dirty := false
```

Comment:
Nit: You might get some better performance out of pre-creating the binding that you want and then issuing an update right way if any field is determined as invalid.

I'm marking this as a nit since although several of these fields (subjects/ownerRefs) are slices of complex objects, in practice they will only contain only a few entries, and I think `reflect.DeepEqual` checks the length so I don't think this will end up having a significant performance impact.

Reply:
I agree that this is probably unnecessary. If it is kept, can the comparison logic be set in its own function? I think this would improve tests as well since the tests' main concern is that logic.

Reply:
Pre-creating a binding and passing it to `Update` has the disadvantage of potentially changing present and future fields I am not interested into, so my thinking was to avoid that. However, I do see your objection that once dirty is set to true, any further checks are wasted.

I also agree it will probably not make a big performance difference, still, in the spirit of trying to set a good example I reworked the logic to enable short-circuiting of checks.

Please let me know if it is acceptable as of e63a011a7.

Reply:
Re-worked logic LGTM, I like the use of the or statement to effectively short-circuti.